src/adapter/src/coord.rs

```diff
@@ -1675,6 +1608,97 @@ impl Coordinator {
         self.apply_local_write(register_ts).await;
     }
 
+    fn bootstrap_pre_optimize(
+        &mut self,
+        ordered_catalog_entries: &[CatalogEntry],
+    ) -> Result<PreOptimizedObjects, AdapterError> {
+        let mut pre_optimized = PreOptimizedObjects::default();
+
+        // The optimizer expects to be able to query its `ComputeInstanceSnapshot` for
+        // collections the current dataflow can depend on. But since we don't yet install anything
+        // on compute instances, the snapshot information is incomplete. We fix that by manually
+        // patching `ComputeInstanceSnapshot` objects to ensure they contain collections previously
+        // optimized.
+        let mut instance_snapshots = BTreeMap::new();
+
+        let optimizer_config = OptimizerConfig::from(self.catalog().system_config());
+
+        for entry in ordered_catalog_entries {
+            let id = entry.id();
+            match entry.item() {
+                CatalogItem::Index(idx) => {
+                    // Collect optimizer parameters.
+                    let compute_instance =
+                        instance_snapshots.entry(idx.cluster_id).or_insert_with(|| {
+                            self.instance_snapshot(idx.cluster_id)
+                                .expect("compute instance exists")
+                        });
+
+                    // Compute instances may have special introspection indexes that get
+                    // automatically installed, so we should ignore them here.
+                    if compute_instance.contains_collection(&id) {
+                        continue;
+                    }
+
+                    // Build an optimizer for this INDEX.
+                    let mut optimizer = optimize::OptimizeIndex::new(
+                        self.owned_catalog(),
+                        compute_instance.clone(),
+                        id,
+                        optimizer_config.clone(),
+                    );
+
+                    // MIR ⇒ MIR optimization (global)
+                    let index_plan = optimize::Index::new(entry.name(), &idx.on, &idx.keys);
+                    let global_mir_plan = optimizer.optimize(index_plan)?;
+
+                    pre_optimized
+                        .indexes
+                        .insert(id, (optimizer, global_mir_plan));
+
+                    compute_instance.insert_collection(id);
```

Comment:
It's a bit ugly that we modify the snapshot here. Is there a better way of ensuring the optimizer knows about previous indexes?